app/models/location.rb

```diff
@@ -0,0 +1,68 @@
+class Location
+  def initialize(city, province, country)
+    @city = city
+    @province = province
+    @country = country
+  end
+
+  attr_reader :city, :province, :country
+
+  # Listed in the order of significance.
+  # The name "component" is pulled from USPS
+  # https://pe.usps.com/text/pub28/28c2_012.htm#ep526349
+  COMPONENTS = [:city, :province, :country]
+
+  def component(compon)
+    send(compon) if compon.in? COMPONENTS
+  end
+
+  def components
+    [@city, @province, @country]
+  end
+
+  def sig_component
+    # The significant component represents the most specific location attribute
+    # that is provided (similar to sig figs in math).
+    COMPONENTS.each do |compon|
+      return compon if send(compon).present?
+    end
+  end
+
+  def ==(other)
+    components == other.components
+  end
+
+  def covers?(other)
+    # Location A covers location B if location A is the more broad than location B.
+
+    return false unless sig_component_value && other.send(:sig_component_value)
+    return false if sig_component_value >= other.send(:sig_component_value) # Location A is more specific than location B.
+
+    sig = other.send(:sig_component_value)
+    COMPONENTS[sig..].all? do |compon|
+      component(compon) == other.component(compon)
+    end
+  end
+
+  def covers_or_equal?(other)
+    self == other || covers?(other)
+  end
+
+  def covered_by?(other)
+    other.covers?(self)
+  end
+
+  def covered_by_or_equal?(other)
+    self == other || covered_by?(other)
+  end
+
+  def to_s
+    components.compact.join(", ")
+  end
+
+  private
```

Comment:
is this is `protected` we can avoid the `send`s

Reply:
ah good idea. please change it to `protected` and remove the `send`s

Reply:
[`9fdb355` (#73)](https://github.com/hackclub/hackathons-backend/pull/73/commits/9fdb3552feca4eb637daf60bfeb65d638e0ab4a7)